File: app/src/main/java/com/junting/drug_android_frontend/services/BluetoothSocket.java
package com.junting.drug_android_frontend.services


import android.app.Activity;
import android.bluetooth.BluetoothAdapter;
import android.bluetooth.BluetoothDevice;
import android.bluetooth.BluetoothSocket;

import java.io.IOException;
import java.io.OutputStream;

public class BluetoothSocket extends Activity {

    private BluetoothSocket mmSocket;
    private OutputStream mmOutputStream;


    // calling this function to send the message to the device
    public static void sendStringOverBluetooth(String message) {
        try {
            BluetoothAdapter bluetoothAdapter = BluetoothAdapter.getDefaultAdapter();
            BluetoothDevice device = bluetoothAdapter.getRemoteDevice("78:21:84:8C:A6:92"); //MAC address of the device

            SerialSocket ss = new SerialSocket(device);
            ss.run();

            byte[] messageBytes = message.getBytes();
            ss.write(messageBytes);

        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public void openPillbox(int position){
        sendStringOverBluetooth(String.format(position));
    }

    public void closePillbox(int position){
        sendStringOverBluetooth(String.format(position));
    }

}

File: app/src/main/java/com/junting/drug_android_frontend/services/BTServices/BluetoothSocket.kt
package com.junting.drug_android_frontend.services.BTServices

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.io.IOException

class BluetoothSocket {

    val ss = SerialSocket()
    init {
        GlobalScope.launch(Dispatchers.IO) {
            ss.run()
        }
    }

    suspend fun openPillbox(cellPosition: String) = withContext(Dispatchers.IO) {
        val message = when (cellPosition) {
            "7" -> "b"
            "8" -> "d"
            "9" -> "f"
            "4" -> "h"
            "5" -> "j"
            "6" -> "l"
            "1" -> "n"
            "2" -> "p"
            "3" -> "r"
            else -> ""
        }
        try {
            val messageBytes = message.toByteArray()
            ss.write(messageBytes)
        } catch (e: IOException) {
            e.printStackTrace()
//            throw e  // 重新拋出異常
        }
    }

    suspend fun closePillbox(cellPosition: String) = withContext(Dispatchers.IO) {
        val message = when (cellPosition) {
            "7" -> "a"
            "8" -> "c"
            "9" -> "e"
            "4" -> "g"
            "5" -> "i"
            "6" -> "k"
            "1" -> "m"
            "2" -> "o"
            "3" -> "q"
            else -> ""
        }
        try {
            val messageBytes = message.toByteArray()
            ss.write(messageBytes)
            ss.close()
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    suspend fun closeMultiplePillbox(arr: IntArray) = withContext(Dispatchers.IO) {
        for (i in arr) {
            val position = i.toString()
            val message = when (position) {
                "7" -> "a"
                "8" -> "c"
                "9" -> "e"
                "4" -> "g"
                "5" -> "i"
                "6" -> "k"
                "1" -> "m"
                "2" -> "o"
                "3" -> "q"
                else -> ""
            }
            try {
                val messageBytes = message.toByteArray()
                ss.write(messageBytes)
            } catch (e: IOException) {
                e.printStackTrace()
            }
        }
        ss.close()
    }
}
